_posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown

```diff
@@ -0,0 +1,372 @@
+
+{% highlight javascript %}
+type Action = {
+  type: string,
+  payload?: any,
+  error?: boolean,
+  meta?: any,
+};
+{% endhighlight %}
+
+Granted, FSA is actually fairly weak in its type constraints, so you could enforce this more strictly as it suits you and your team. Still, this helps when creating your Flux modules:
+
+{% highlight javascript %}
+export const completeTodo = (id: string): Action => ({
+  type: TODO_COMPLETE, // constant declared above
+  payload: {
+    id,
+  },
+})
+{% endhighlight %}
+
+If you were to return an object that didn't adhere to the `Action` type Flow would learn about it, hunt you down, and <s>kill you</s> fail your CI build.
+
+### State Shapes
+
+The obvious benefit here is that we can create a type alias for each reducer's shape, allowing Flow to analyze that each action handler in a reducer produces a conforming state shape. This is fairly trivial once you've annotated a function or two, since reducers are _just plain functions_, but there's an even bigger benefit to be had.
+
+Say we have a reducer responsible for some search state, perhaps a form with a handful of fields, nothing crazy. When the filters are submitted (in this example), we update the store to cache that state for returning visitors to provide some decent UX. This is super high-level, so please forgive me for not going through it in its entirety.
+
+{% highlight javascript %}
+// your filter state shape (presumably in a redux module)
+export type MySearchFilterState = {
+  selectedUserIds: Array<string>,
+  searchFromDate: number,
+  searchToDate: number,
+  hasSomeFilterEnabled: boolean,
+};
+
+// consumer component (parent smart component):
+import type { MySearchFilterState } from 'my/redux/module'
+
+class MySmartViewOrSomething extends React.Component {
+  _handleSubmitFilters(filters: MySearchFilterState): void {
+    // dispatch an action to cache these filters
+  }
+
+  render(): React.Element {
+    return (
+      // a bunch of stuff...
+      <MyFormComponent onSubmit={this._handleSubmitFilters} />
+    )
+  }
+}
+
+// consumer component's child
+type Props = {
+  filters: MySearchFilterState,
+  onSubmit: Function,
+}
+class MyFormComponent extends React.Component {
+  // ...
+  _handleSubmitFilters(): void {
+    this.props.onSubmit({
+      selectedUserIds: [...],
+
+      // "The Witch" is pretty good, you should see it.
+      someBadProperty: 'Black Phillip',
+    })
+  }
+  // ...
+}
+{% endhighlight %}
+
+The amazing part about this setup is that if `<MyFormComponent />` decides to not abide by the shape of the filters (as it fails to do above), Flow will catch it without ever having to run the application. Whether the error is an additional/malformed property, a missing required property, or an incorrect type, Flow has you covered. If you ever change the shape your state, Flow will help you confirm that the rest the application continues humming along.
```

Comment:
You may want to reinforce your point here about this kind of error leaking into production without flow.